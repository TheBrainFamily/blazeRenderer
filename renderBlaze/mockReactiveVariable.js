class _ReactiveVar {
  constructor(value) {
    this.value = value;
  }
  set(value) {
    this.value = value;
  }
  get() {
    return this.value;
  }
}

ReactiveVar = _ReactiveVar