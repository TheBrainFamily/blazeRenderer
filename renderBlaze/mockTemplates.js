import _ from 'underscore'

var handler = {
  get: function (target, name) {
    if (name === 'registerHelper') {
      return function(functionName, functionPassed) {
        if(!target.globalHelpers) {
          target.globalHelpers = {};
        }
        if(_.isFunction(functionPassed)) {
          target.globalHelpers[functionName] = function () {
            return functionPassed.apply(this, arguments);
          }
        }
      }
    }
    if (name === 'instance') {
      return function () {
        return target[window.CurrentTemplate]
      }
    }
    if (name === 'subscriptionsReady') {
      return function() {
        return true;
      }
    }
      if (!name || !target || !(name in target)) {
        target[name] = {
          helpers: function (helpers) {
            this.helpers = helpers
          },
          getHelpers: function () {
            function wrapperFunction(key) {
              const targetObject = helpers[key] ? helpers : target.globalHelpers;
              if (!_.isFunction(targetObject[key])) {
                wrappedHelpers[key] = targetObject[key];
              }
              else {
                wrappedHelpers[key] = function () {
                  if (!window.PreviousTemplate || window.PreviousTemplate !== name) {
                    window.PreviousTemplate = window.CurrentTemplate
                    window.CurrentTemplate = name
                  }
                  const value = _.isFunction(targetObject[key]) ? targetObject[key].apply(this, arguments) : targetObject[key]
                  window.CurrentTemplate = window.PreviousTemplate
                  return value;
                }
              }
            }
            const wrappedHelpers = {}
            const helpers = this.helpers
            if(target.globalHelpers) {
              Object.keys(target.globalHelpers).forEach(wrapperFunction);
            }
            Object.keys(helpers).forEach(wrapperFunction);

            return wrappedHelpers
          },
          onCreated: function (callback) {
            this._onCreatedCallback = callback
          },
          _runOnCreatedCallback() {
            this._onCreatedCallback && this._onCreatedCallback.apply(this)
          },
          onRendered() {},
          onDestroyed() {},
          events() {},
          //TODO this most probably shouldn't be here but attached to the callbacks like onCreated
          subscribe() {},
          autorun: function(callback) {
            //TODO not sure if we should bind this...
            callback.apply(this)
          },
        }
      }
      return target[name]
    }
}

Template = new Proxy({}, handler)

