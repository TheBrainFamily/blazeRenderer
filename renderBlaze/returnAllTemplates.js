import glob from 'glob'

export default function(path) {

  const value = [path].map(function (path) {
    console.log("Gandecki process.cwd()", process.cwd());
    if (path.match(/.*\*\*\/\*.html/)) {
      console.log("Gandecki glob.sync(path, {cwd: process.cwd()})", glob.sync(path, {cwd: process.cwd()}));
      return glob.sync(path, {cwd: process.cwd()})
    }

    if (path.match(/.*\.html/)) {
      console.log("Gandecki path", path);
      return path
    }
    console.log("Gandecki glob.sync(path + '/**/*.html', {cwd: process.cwd()})", glob.sync(path + '/**/*.html', {cwd: process.cwd()}));
    return glob.sync(path + '/**/*.html', {cwd: process.cwd()})
  }).reduce(function (a, b) {
    return a.concat(b)
  }, [])

  console.log("Gandecki value", value);
  return value;
}