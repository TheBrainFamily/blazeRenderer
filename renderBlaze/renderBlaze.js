import html from './blazeInternals/html'
import compile from './blazeInternals/compiler'
import fs from 'fs';
import { html as beautifyHtml } from 'js-beautify'
import './blazeInternals/blazeWith'
import './mockTemplates'
import './mockReactiveVariable'

var toHTML = function (data, template) {
    var compiled = compile(template, {isBody: true});
    var fn = eval(compiled);
    return BlazeMine.toHTML(BlazeMine.With(data, fn));
};

export const renderBlazeWithData = function renderBlaze(templateFile, templateName, data) {
    const include = function includeReplacement(templateName, data) {
        data = data || {};
        data = Object.assign({}, data, {includeReplacement})

        return toHTML(data, $(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"'));
    }

    const template = fs.readFileSync(templateFile)
    var cheerio = require('cheerio');
    var $ = cheerio.load(template.toString().replace(/({{> *)([^\s]*) }}/g, '{{{ includeReplacement \'$2\' }}}'));

    return beautifyHtml(include(templateName, data))
}


const renderBlaze = function (templateName) {

}

export default renderBlaze;
export const parseTemplates = function (templateFiles) {
    const templatesToFilesMap = []
    templateFiles.forEach((templateFile) => {
        const template = fs.readFileSync(templateFile)
        var cheerio = require('cheerio');
        $ = cheerio.load(template.toString().replace(/({{> *)([^\s]*) }}/g, '{{{ includeReplacement \'$2\' }}}'));
        $('template').each((index, foundTemplate) => {
            templatesToFilesMap.push({templateName: $(foundTemplate).attr('name'), templateFile, cheerio: $})
        })
    })
    return templatesToFilesMap;
}

export const renderBlazeWithTemplates = function (templateName, parsedTemplates) {
    const includeReplacement = function includeReplacement(templateName) {
        let data = Template[templateName].getHelpers() || {};
        data = Object.assign({}, data, {includeReplacement})
        let cheerio;
        cheerio = parsedTemplates.find(template => template.templateName === templateName).cheerio

        let template = cheerio(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"');
        return toHTML(data, template);
    }
    return includeReplacement(templateName)
}


