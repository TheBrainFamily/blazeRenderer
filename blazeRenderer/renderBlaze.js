import html from './blazeInternals/html'
import compile from './blazeInternals/compiler'
import fs from 'fs';
import { html as beautifyHtml } from 'js-beautify'
import './blazeInternals/blazeWith'
import './blazeInternals/template'
import './mockTemplates'
import './mockReactiveVariable'
import returnAllTemplates from './returnAllTemplates'

var toHTML = function (data, template, templateName, matchedInsideTemplates) {

    if (matchedInsideTemplates) {
      matchedInsideTemplates.forEach(insideTemplate => {
        var compiledInside = compile(insideTemplate.templateInside ,{isBody: true})
        var fnInside = eval(compiledInside)
        window.hackedInTemplates = {}
        window.hackedInTemplates[insideTemplate.name] = fnInside
      })
    }

    var compiled = compile(template, {isBody: true});
    var fn = eval(compiled)
    return BlazeMine.toHTML(BlazeMine.With(data, fn, templateName));
};

export const renderBlazeWithData = function renderBlaze(templateFile, templateName, data) {
    const include = function includeReplacement(templateName, data) {
        data = data || {};
        data = Object.assign({}, data, {includeReplacement})

		return toHTML(data, $(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"'), templateName);
    }

    const template = fs.readFileSync(templateFile)
    var cheerio = require('cheerio');
    var $ = cheerio.load(template.toString().replace(/{{> *([^\s}]*)([^}]*)}}/g, '{{{ includeReplacement \'$1\' $2 }}}'));

    return beautifyHtml(include(templateName, data))
}

export const parseTemplates = function (templateFiles) {
  const templatesToFilesMap = []
  templateFiles.forEach((templateFile) => {
    const template = fs.readFileSync(templateFile)
    //TODO add test cases for multiline {{> }}
    //TODO add test case for a case when {{> were}} <- no space
    const parsedText = template.toString().replace(/{{(>) ?(Template.contentBlock)/g, '{{ $2').replace(/{{> *([^\s}]*)([^}]*)}}/g, '{{{ includeReplacement \'$1\' $2 }}}').replace(/({{|{{.*( |=))(this)( |}})/g, '$1_myOwnThis$4').replace(/&gt;/g, '>').replace(/&apos;/g, '\'').replace(/&quot;/g, '"')

    const templateRegex = /<template name=("|')(.*)("|')>((.|\n)*?)<\/template>/gm

    let match = templateRegex.exec(parsedText)

    while (match !== null) {
      templatesToFilesMap.push({templateName: match[2].trim(), templateFile, cheerio: match[4].trim()})
      match = templateRegex.exec(parsedText)
    }

  })
    return templatesToFilesMap;
}

const renderBlazeWithTemplates = function (templateName, parsedTemplates) {
  if (parsedTemplates && parsedTemplates.length > 0) {

  } else {
    parsedTemplates = parseTemplates(returnAllTemplates('imports/').concat(returnAllTemplates('client/')))
  }
    const includeReplacement = function includeReplacement(templateName) {
    let passedArguments;
    if (Array.from(arguments)[1] ) {
      passedArguments = Array.from(arguments)[1]['hash']  ? Array.from(arguments)[1]['hash'] : {_myOwnData: Array.from(arguments)[1]}
    }
        Template[templateName].helpers = Object.assign({}, Template[templateName].getHelpers(), passedArguments, {isInRole: function() { return true }}, {$or: function(arg1, arg2) { return arg1 || arg2}}, {$gt: function(arg1, arg2) { return arg1 > arg2}}, {$eq: function(arg1, arg2) { return arg1 === arg2 }}, {$exists: function(arg1) { return !!arg1 }}, {pathFor: function(arg1, arg2) { return `${arg1}/${arg2}`}}, {_myOwnThis: function() {return this._myOwnData}}, {$:{Session:{get: function(arg) { return true }}}})

      //TODO add test for isInRole, and most probably make this configurable instead of hardcoded.
      // Used in https://github.com/alanning/meteor-roles
      data = Object.assign({}, Template[templateName].getHelpers(), {includeReplacement})

      const cheerioPotentially = parsedTemplates.find(template => template.templateName === templateName)

      let template;
      if (!cheerioPotentially) {
        console.log("Gandecki templateName", templateName);
        template = '<div></div>'
      } else {
        template = cheerioPotentially.cheerio
      }
      // console.log(`${templateName} - ${template}`);

      let myRegexp = /{{ ?#(?!.*if|.*unless|.*each|.*with)([^ }]*).*/g

      let match = myRegexp.exec(template);

      let matchedInsideTemplateNames = []

      while (match !== null) {
        // matched text: match[0]
        // match start: match.index
        // capturing group n: match[n]
        matchedInsideTemplateNames.push(match[1].trim())
        match = myRegexp.exec(template);
      }


      let matchedInsideTemplates = matchedInsideTemplateNames.map(name => {
        let cheerioInside = parsedTemplates.find(template => template.templateName === name).cheerio
        let templateInside = cheerioInside
        return {templateInside, name}
      })


        return toHTML(data, template, templateName, matchedInsideTemplates);
    }
    return beautifyHtml(includeReplacement(templateName))
}

export default renderBlazeWithTemplates;
