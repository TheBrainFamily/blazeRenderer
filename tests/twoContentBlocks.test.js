import renderBlaze, { renderBlazeWithData, parseTemplates } from '../blazeRenderer/renderBlaze';

describe('TwoContentBlocks', function () {
  describe('templates', function () {
    it('should be able to render two or more templates with contentBlock inside', function () {
      expect(renderBlaze('wrapper')).toMatchSnapshot()
    })
  })
})
