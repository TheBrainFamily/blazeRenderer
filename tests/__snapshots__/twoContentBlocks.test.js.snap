// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoContentBlocks templates should be able to render two or more templates with contentBlock inside 1`] = `
"<span class=\\"passedParamTable\\">paramForTable</span>

<span>Inside table</span>

<span class=\\"passedParamModal\\">paramForModal</span>

<span>Inside modal</span>"
`;
